src/NuGet.Clients/NuGet.VisualStudio.Common/OutputConsoleLogger.cs

```diff
@@ -30,6 +33,9 @@ public class OutputConsoleLogger : INuGetUILogger, IDisposable
         private EnvDTE.BuildEvents _buildEvents;
         private EnvDTE.SolutionEvents _solutionEvents;
 
+        [SuppressMessage("Build", "CA2213:'OutputConsoleLogger' contains field '_semaphore' that is of IDisposable type 'ReentrantSemaphore', but it is never disposed. Change the Dispose method on 'OutputConsoleLogger' to call Close or Dispose on this field.", Justification = "Field is disposed from async task invoked from Dispose.")]
+        private readonly ReentrantSemaphore _semaphore = ReentrantSemaphore.Create(0, NuGetUIThreadHelper.JoinableTaskFactory.Context, ReentrantSemaphore.ReentrancyMode.NotAllowed);
```

Comment:
You're not handing in a `JoinableTaskFactory` instance at all here -- only the singleton `JoinableTaskContext` which is the same as `ThreadHelper.JoinableTaskContext`. So no, the tasks queued on this semaphore will _not_ automatically block shutdown. 

To ensure your fire and forget tasks are completed before your AsyncPackage is disposed of, you'll need to either use JTF.RunAsync around each of your ExecuteAsync calls in order to associate it with your AsyncPackage, or we might expose a way from the `ReentrantSemaphore` class itself for you to "block" on completion of all the items currently in the queue.

Reply:
I didn't see a constructor accepting JTF?

Reply:
There isn't one.

Reply:
Wrapped into JTF.RunAsync.